### PR TITLE
Update Crowdstrike pagination when more then 10000 hosts

### PR DESCRIPTION
**Pull request by LarsenEric**

### CODE REVIEW

File: connectors/crowdstrike/connector/server_access.py [hunk 1]
```diff
@@ -60,7 +60,7 @@ def get_hosts(self, last_model_state_id=None):
                     for err in error_detail:
                         raise DatasourceFailure(err["message"], err["code"])
                 if len(hosts) < host_lookup['body']['meta']['pagination']['total']:
-                    offset = len(hosts)
+                    offset = host_lookup['body']['meta']['pagination']['offset']
```
Comment (mdazam1942) [inline connectors/crowdstrike/connector/server_access.py]: use python dictionary get() method to avoid the key errors. https://docs.python.org/3/library/stdtypes.html#dict.get
since we are unsure about the offset value we need to use this method to get the correct value and handle the exception as well.

File: connectors/crowdstrike/connector/server_access.py [hunk 2]
```diff
@@ -59,13 +62,36 @@ def get_hosts(self, last_model_state_id=None):
                     error_detail = host_lookup["body"]["errors"]
                     for err in error_detail:
                         raise DatasourceFailure(err["message"], err["code"])
-                if len(hosts) < host_lookup['body']['meta']['pagination']['total']:
-                    offset = len(hosts)
-                    host_lookup = discover.query_hosts(filter=result_filter, limit=DISCOVER_PAGE_SIZE, offset=offset)
+                hosts = remove_duplicate_hosts(hosts)
```
Comment (DerekRushton) [inline connectors/crowdstrike/connector/server_access.py]: Why is this needed? Won't it clean it up at the end?

File: connectors/crowdstrike/connector/server_access.py [hunk 3]
```diff
@@ -59,13 +62,36 @@ def get_hosts(self, last_model_state_id=None):
...
+                        last_added_filter = "last_seen_timestamp:>='" + last_added + "'"
+                        if result_filter == "":
+                            updated_filter = last_added_filter
+                        else:
+                            updated_filter = result_filter + " + " + last_added_filter
+
+                        host_lookup = discover.query_hosts(filter=updated_filter, limit=DISCOVER_PAGE_SIZE, sort=sort)
+                    else:
+                        offset = len(hosts)
+                        host_lookup = discover.query_hosts(filter=result_filter, limit=DISCOVER_PAGE_SIZE, offset=offset, sort=sort)
```
Comment (DerekRushton) [inline connectors/crowdstrike/connector/server_access.py:89]: Does this line still use the original query instead of the updated filter? Would that cause problems when going for 10000 - 10100 where it's still using the original query and getting results 0-100 instead of results 10,000-10100?

File: connectors/crowdstrike/connector/server_access.py (hunk 2 above)
Comment (LarsenEric) [reply connectors/crowdstrike/connector/server_access.py]: The problem is the next line.
If we have duplicates at this point in the hosts list then there is a chance we do not get all of the hosts.
For example:
We know the total.
We go through and in the process of ensuring we get all the hosts we have a couple duplicates then when it checks the length of the hosts against the total it may seem like we got them all and will skip trying to get any more.
Let me know if that makes sense.

File: connectors/crowdstrike/connector/server_access.py (hunk 3 above)
Comment (LarsenEric) [reply connectors/crowdstrike/connector/server_access.py:89]: Yeah, I think there is a case I missed here. Will take a look.

Comment (LarsenEric) [reply connectors/crowdstrike/connector/server_access.py:89]: This should be resolved by just updating the result_filter variable instead of creating a new variable.
Will quickly test this

File: connectors/crowdstrike/connector/server_access.py (hunk 2 above)
Comment (DerekRushton) [reply connectors/crowdstrike/connector/server_access.py]: This may require a bit of a logic change. I'm not a fan of deduplicating on every loop. 

Instead of using the total number of results, it may be better to check only the current query. 

What I mean by this is that we know that each page will have 100 results (provided there is more to get). If the results is less than 100, than you know you are done. This would only be the exit condition for the loop. The other logic would be untouched.

Let me know if this is viable.

Comment (LarsenEric) [reply connectors/crowdstrike/connector/server_access.py]: Will take a look

File: connectors/crowdstrike/connector/data_handler.py [hunk 4]
```diff
@@ -52,6 +52,9 @@ def group_host_sensor_apps(applications):
             group_apps[agent_id].append(app)
     return group_apps
 
+def remove_duplicates(items):
+    """Remove any duplicate hosts from the array of hosts"""
```
Comment (DerekRushton) [inline connectors/crowdstrike/connector/data_handler.py:57]: Would be nice to have some logging on this one. I'd log it as "debug" as it would be very chatty output.

File: connectors/crowdstrike/connector/server_access.py [hunk 5]
```diff
@@ -42,13 +42,17 @@ def get_hosts(self, last_model_state_id=None):
...
         hosts = []
         try:
             discover = Discover(client_id=self.client_id, client_secret=self.client_secret, base_url=self.base_url)
+            offset = 0
             result_filter = ""
+            sort = "last_seen_timestamp|asc"
             if last_model_state_id:
                 result_filter = f"last_seen_timestamp:>'{epoch_to_datetime_conv(last_model_state_id)}'"
-            host_lookup = discover.query_hosts(filter=result_filter, limit=DISCOVER_PAGE_SIZE)
+            host_lookup = discover.query_hosts(filter=result_filter, limit=DISCOVER_PAGE_SIZE, sort=sort)
```
Comment (DerekRushton) [inline connectors/crowdstrike/connector/server_access.py:55]: Lets have a log here to keep track of some meta data about the queries we are making.

We should include a "Starting Host Query" before and a "Successful Host Query" line after this call.

Comment (DerekRushton) [reply connectors/crowdstrike/connector/server_access.py:55]: For the after, we can include the "total count" as well. Just make sure that "total count" actually exist before posting it that way.

Comment (DerekRushton) [reply connectors/crowdstrike/connector/server_access.py:55]: The filter shouldn't be changed at this point either. We could post that in the "starting query"

File: connectors/crowdstrike/connector/server_access.py [hunk 6]
```diff
@@ -59,13 +63,37 @@ def get_hosts(self, last_model_state_id=None):
...
+                        last_added = deep_get(hosts[-1], ['last_seen_timestamp'])
+                        last_added_filter = "last_seen_timestamp:>='" + last_added + "'"
+                        if result_filter == "":
+                            updated_filter = last_added_filter
+                        else:
+                            updated_filter = result_filter + " + " + last_added_filter
+
+                        host_lookup = discover.query_hosts(filter=updated_filter, limit=DISCOVER_PAGE_SIZE, sort=sort)
+                    else:
+                        offset = len(hosts)
```
Comment (DerekRushton) [inline connectors/crowdstrike/connector/server_access.py:87]: Lets add a debug log here with 

"Querying host with offset = X." We already know the filter. But feel free to include it in each debug if you want.

Comment (DerekRushton) [reply connectors/crowdstrike/connector/server_access.py:87]: Also include a "successful" type of message after.

File: connectors/crowdstrike/connector/server_access.py [hunk 7]
```diff
@@ -59,13 +63,37 @@ def get_hosts(self, last_model_state_id=None):
...
+                if len(host_detail["body"]["resources"]) == DISCOVER_PAGE_SIZE:
+                    '''CrowdStrike's Discover endpoint has the following limitation: limit + offset <= 10,000
+                        For that reason we need to do additional queries to get all hosts. We do this by making
+                        additional queries and filter on the last_seen_timestamp. This also means we have the
+                        potential of collecting duplicate hosts. So we make sure to remove these duplicates.'''
+                    if DISCOVER_PAGE_SIZE + offset > 10000:
+                        offset = 0
+                        last_added = deep_get(hosts[-1], ['last_seen_timestamp'])
+                        last_added_filter = "last_seen_timestamp:>='" + last_added + "'"
+                        if result_filter == "":
```
Comment (DerekRushton) [inline connectors/crowdstrike/connector/server_access.py:78]: Lets add a debug log here with the new result filter and offset count.

File: connectors/crowdstrike/connector/server_access.py [hunk 8]
```diff
@@ -59,13 +63,37 @@ def get_hosts(self, last_model_state_id=None):
...
+                            updated_filter = last_added_filter
+                        else:
+                            updated_filter = result_filter + " + " + last_added_filter
+
+                        host_lookup = discover.query_hosts(filter=updated_filter, limit=DISCOVER_PAGE_SIZE, sort=sort)
```
Comment (DerekRushton) [inline connectors/crowdstrike/connector/server_access.py:84]: And once more a debug line here for each successful call.

File: connectors/crowdstrike/connector/server_access.py [hunk 9]
```diff
@@ -59,13 +63,37 @@ def get_hosts(self, last_model_state_id=None):
...
+                hosts = remove_duplicates(hosts)
+            
         except Exception as ex:
-            raise DatasourceFailure(ex)
+            raise DatasourceFailure(ex)      
+
+        context().logger.info("Imported %s hosts from the CrowdStrike Falcon environment.", len(hosts))
         return hosts
 
     def get_applications(self, app_filter=None, last_model_state_id=None):
```
Comment (DerekRushton) [inline connectors/crowdstrike/connector/server_access.py:105]: For the remaining rest calls can you do something similar for debugging. This way we can keep track of what calls are being made and what the parameters of those calls are?

**Review by DerekRushton — APPROVED** [review]

**Review by DerekRushton — APPROVED** [review]
